### ISSUE TITLE
apply_transform function throwing exception

### ISSUE THREAD

**Issue by Kanyq** [voxelmorph/voxelmorph#300]
Hello,

I'm trying to warp my moving segmentation image according to the learned warping using the "apply_transform" function of vxm_model but it is throwing me an exception saying "Offset flow field size expected: 1, found: 3".

transformed_image = vxm_model.apply_transform(src, trg, img, interp_method='nearest')

src: the original moving image (256,256,64)
trg: moved predicted image (256,256,64)
img: segmentation image to be warped (256,256,64)

Any explanation of the function parameters or the cause of the exception is really appreciated. As the "apply_transform" code doesn't go in details about the parameters I'm not sure if I'm passing the correct parameters either.

**Comment by adalca** [comment]
the `src`, `trg`, and `img` still have to be in keras size -- so in this case, `[1, 256, 256, 64, 1]`

**Comment by Kanyq** [comment]
Hello,
I reshaped my arrays to [1,256,256,64,1] and applied the "apply_transform" function on it but the predicted transformation is all zeroes.
transformed_image = vxm_model.apply_transform(src, trg, img, interp_method='nearest')

src: the original moving image [1,256,256,64,1]
trg: moved predicted image [1,256,256,64,1]
img: segmentation image to be warped [1,256,256,64,1]
transformed_image: [0..0] of shape [1,256,256,64,1]

I'm not sure what could I be doing wrong here. As my predictions for the moved_image seems fine. Thanks for your help.

**Comment by adalca** [comment]
@Kanyq can you please share some screenshots or something like that? Also, how was the model trained?

**Comment by Kanyq** [comment]
The model was trained on 256,256,64 image resolution which was intensity normalized in the range [0,1]. 
for convulational layer: kernel_size = 3, padding = "same"

I have modified the **data generator** to suit my project needs: our data is a numpy array of moving and fix images (moving images are on even indexes like 0,2,4.. and their corresponding fix images on odd indexes like 1,3,5.. respectively).
```
idx1 = np.random.randint(0, x_data.shape[0], size=batch_size)  #batch_size=1
        if idx1%2 == 0:   #for even idx1 we're using it as moving image index and idx+1 as fixed image index
            moving_images = x_data[idx1, ..., np.newaxis]
            fixed_images = x_data[idx1+1, ..., np.newaxis]
        else:   #for odd idx1 we're using it as fix image index and idx-1 as moving image index
            moving_images = x_data[idx1-1, ..., np.newaxis]
            fixed_images = x_data[idx1, ..., np.newaxis]
inputs = [moving_images, fixed_images]

```
For "**vxm_model**": vxm_model = vxm.networks.VxmDense(vol_shape, nb_features, int_steps=0)
vol_shape: 256,256,64
nb_features = [
    [16, 32 , 32, 32],         # encoder features
    [32, 32 , 32, 32, 32, 16, 16]  # decoder features
]

**Losses:**
losses = ['mse', vxm.losses.Grad('l2').loss]
loss_weights = [1, 0.01]

**Optimizer:**
optimizer=tf.keras.optimizers.Adam(lr=1e-3)

**Training:**
epochs: 30 epochs
steps_per_epochs: 88 (There are 88 training images, 44 pairs and since I'm using a batch_size of 1, I went with 88 SPE)

**Training loss plot:** Final loss: 0.271
![loss](https://user-images.githubusercontent.com/14558146/116408686-666ab600-a833-11eb-8b94-95537b3fa0d6.png)

**Model Prediction:**
![prediction](https://user-images.githubusercontent.com/14558146/116408797-8306ee00-a833-11eb-90b8-26b0f73ae9d4.png)

**apply_transform on segmentation image:** Range[-1,1]
![Screenshot 2021-04-28 at 14 52 56](https://user-images.githubusercontent.com/14558146/116408932-a5007080-a833-11eb-914a-e51a1c1883b4.png)

**Segmentation warping:**
![seg](https://user-images.githubusercontent.com/14558146/116409752-7040e900-a834-11eb-8469-c95059ead822.png)

The segmentation warping was generated using:
warped_seg_mov = warp_model.predict([segmentation_array[17][np.newaxis,...,np.newaxis], pred_warp])
segmentation_array[17]: moving segmentation image of size 256,256,64
pred_warp: 
val_pred = vxm_model.predict(val_input)
moved_pred = val_pred[0].squeeze() #This is the predicted moving image
pred_warp = val_pred[1] #This is used as warping for segmentation images

Let me know if you need anymore information from me. Thank you

**Comment by adalca** [comment]
Thanks. A couple of quick question/comments:

- I would recommend just grabbing 2 random images at each iteration, rather than the odd/even scheme you have. It should give you much richer models
- What is `warp_model` exactly in your case? I thought you said you used `apply_transform`?
- Can you try the apply transform route to warp the image as well (not just the segmentation) and see how that comes out?

**Comment by Kanyq** [comment]
- Okay, but if I grab 2 random images it's possible that I grab two fixed or two moving images also it's possible that I grab a different fixed image and a different moving image. Wouldn't that train the model wrong? Since we want to learn transformation from moving to fixed, shouldn't we pick the images in pair?
- warp_model is: warp_model = vxm.networks.Transform(vol_shape, interp_method='nearest')
- For warping the images I'm just plotting the predicted moved image by the model and I'm not sure, what is the difference between trg and img parameter for the apply_transform function. I can imagine that the moving image will be the src and fix image will be the trg but what should be the img in this case?

**Comment by adalca** [comment]
> * Okay, but if I grab 2 random images it's possible that I grab two fixed or two moving images also it's possible that I grab a different fixed image and a different moving image. Wouldn't that train the model wrong? Since we want to learn transformation from moving to fixed, shouldn't we pick the images in pair?

Maybe I'm not sure what your data is, but I'm not sure why you have sectioned them off in 'moving images' and 'fixed images'? Moving and fixed are just terminology used in the model -- but again, maybe it would help to know more about the data.

> * warp_model is: warp_model = vxm.networks.Transform(vol_shape, interp_method='nearest')

Ok, got it

> * For warping the images I'm just plotting the predicted moved image by the model and I'm not sure, what is the difference between trg and img parameter for the apply_transform function. I can imagine that the moving image will be the src and fix image will be the trg but what should be the img in this case?

`img` would be the image you want to warp. So to warp an image you could do `vxm_model.apply_transform(mov, fix, mov)` and to warp a seg you could do `vxm_model.apply_transform(mov, fix, mov_seg, 'nearest')`

**Comment by Kanyq** [comment]
> Maybe I'm not sure what your data is, but I'm not sure why you have sectioned them off in 'moving images' and 'fixed images'? Moving and fixed are just terminology used in the model -- but again, maybe it would help to know more about the data.
>> The data that we've been provided has fixed and moving image pairs (for both image and segmentation). This is how our data looks.
![files](https://user-images.githubusercontent.com/14558146/116416089-560a0980-a83a-11eb-9542-f9d45a43dcb5.png)

> `img` would be the image you want to warp. So to warp an image you could do `vxm_model.apply_transform(mov, fix, mov)` and to warp a seg you could do `vxm_model.apply_transform(mov, fix, mov_seg, 'nearest')`
>> Okay so my understanding of it was correct but as I said when I use "apply_transform" function for segmentation images I'm getting a zero array of (1,256,256,64,1) and for "images" I'm just plotting the prediction of model and not actually applying the "apply_transform" function on moving image to get the moved image. But I'll try and see if "apply_transform" works in the image data or if it gives me a zero array again. Can you suggest a reason for why I'm getting the zero array after applying "apply_transform"?

**Comment by adalca** [comment]
> The data that we've been provided has fixed and moving image pairs (for both image and segmentation). This is how our data looks.

Right, but is there something special about those images? E.g. are they different scans of the same subject? If they're all just different subjects, I think you should just randomize. 

> Okay so my understanding of it was correct but as I said when I use "apply_transform" function for segmentation images I'm getting a zero array of (1,256,256,64,1) and for "images" I'm just plotting the prediction of model and not actually applying the "apply_transform" function on moving image to get the moved image. But I'll try and see if "apply_transform" works in the image data or if it gives me a zero array again. Can you suggest a reason for why I'm getting the zero array after applying "apply_transform"?

I don't see why it would give you 0s at all really. Can you show me the exact line you used?

**Comment by Kanyq** [comment]
> Right, but is there something special about those images? E.g. are they different scans of the same subject? If they're all just different subjects, I think you should just randomize.
>> I think the image pairs are from the same subject but I'll try your suggestion and see if I'm getting better results. 

> I don't see why it would give you 0s at all really. Can you show me the exact line you used?
>>transformed_image = vxm_model.apply_transform(mov_reshape, fix_reshape, segmentation_array_reshaped, interp_method='nearest')
Here, mov_reshape is moving image reshaped to (1,256,256,64,1) shape. "fix_shape" is fixed image reshaped to (1,256,256,64,1) shape and "segmentation_array_reshaped" is segmentation image reshaped to (1,256,256,64,1) shape.

**Comment by adalca** [comment]
Weird, that should work.

Can you do `transformed_image = vxm_model.apply_transform(mov_reshape, fix_reshape, mov_reshape, interp_method='linear')` and post what you get? 

I suspect you have a bug in the image processing or something like that, but I'm just guessing.

**Comment by Kanyq** [comment]
I tried doing transformed_image = vxm_model.apply_transform(mov_reshape, fix_reshape, mov_reshape, interp_method='linear') and I got the predicted moved image from the model.
![prediction](https://user-images.githubusercontent.com/14558146/116433408-c2d8d000-a849-11eb-8ae0-47ade39a064c.png)
![apply_transform](https://user-images.githubusercontent.com/14558146/116433466-d2581900-a849-11eb-8e8e-58e929f597c1.png)

I rerun the apply_transform with segmentation image and now I'm getting a warping which was similar to the initial warping I showed.
![seg](https://user-images.githubusercontent.com/14558146/116433973-4b577080-a84a-11eb-8307-a7608703af60.png)
![warped](https://user-images.githubusercontent.com/14558146/116433995-4f838e00-a84a-11eb-9b8d-432d892bc088.png)

Even though I can generate warping from the apply_transform function but it doesn't match the fixed image that well(it's closer to the moving image). Do you think the model is underfitting and I should train it for higher epochs(maybe 50epochs)?

**Comment by adalca** [comment]
it looks to me like the warp is a bit bad (under regulaized) here. 

To visualize if this is the case, i recommend you get a bwgrid via `pystrum.pynd.ndutils.bw_grid(img.shape, 5)` and warp this with apply_transform. 

If these images are from the same subject (each moving and fixed pair), then perhaps the regularization should be stronger. You might also want to look at our recent paper https://arxiv.org/abs/2101.01035

**Comment by Kanyq** [comment]
I obtained the bw_grid for "fixed img" and kept rest of the parameters same and this is the warping I got:
![warping](https://user-images.githubusercontent.com/14558146/116453691-77c9b780-a85f-11eb-9e45-f0459877563e.png)
![seg](https://user-images.githubusercontent.com/14558146/116453836-9c259400-a85f-11eb-95dc-fb0cfae21fa3.png)

This warped image is close to the fixed segmentation, which is what I wanted.
However, I don't understand the concept of bw_grid, and am I supposed to apply it to moving or fixed image?

**Comment by adalca** [comment]
I meant something *like* 

```
grid = pystrum.pynd.ndutils.bw_grid(mov.shape, 5)[np.newaxis, ..., np.newaxis]
transformed_image = vxm_model.apply_transform(mov_reshape, fix_reshape, grid , interp_method='linear')
```

The idea is to visualize how `grid` is warped. See Fig 6 in https://arxiv.org/pdf/1809.05231.pdf

**Comment by Kanyq** [comment]
Okay I understand, here is how the warped grid looks like:
![w_grid](https://user-images.githubusercontent.com/14558146/116459496-2c66d780-a866-11eb-81b8-39be698211a1.png)

**Comment by adalca** [comment]
Right, it seems a bit underregularize (not very smooth), likely because you used the same regularization parameter that we used to register brains from different people. If mov and fix are from the same person, it's likely you'd want stronger regularization.

**Comment by Kanyq** [comment]
Okay, so I've re-trained the model and have adjusted the loss weights and set int_steps=4.
For losses, I'm wnet with the implemented loss functions `losses = ['mse', vxm.losses.Grad('l2').loss]`
But I've adjusted the weights from `loss_weights = [1, 0.01]` to `loss_weights = [1, 0.05]` and I'm still not getting satisfactory results with segmentation images. Is there any other regularization that I should try or should I amplify(something like this maybe loss_weights = [1, 0.10]) this regularization a little more?

**Comment by adalca** [comment]
If it's still too rough, yes! This is why we wrote hypermorph, btw.
